### ISSUE TITLE
New media browser: Consolidate attributes

### ISSUE THREAD

**Issue by libardo** [n2cms/n2cms#675]
These editables: 
* EditableFileUploadAttribute
* EditableMediaUploadAttribute
* EditableImageUploadAttribute
* EditableImageAttribute

Should all use the media browser from:
* EditableMediaAttribute

**Comment by DejanMilicic** [comment]
Additionally, it would be great to introduce properties on EditableMediaAttribute that would enable specification of allowed extensions, roles and maybe few more things

**Comment by SntsDev** [comment]
Yes, I think all of them should use the new **EditableMediaUploadAttribute** as well. As this being my first "big" add-on to the project, I knew I shouldn't touch the things that already work. But of course, I won't use the other editables anymore.

What to do? I think that this could work:
* Inherit the other Editables from the new one
* Mark as "obsolete" all the previous
Opinions?

Regarding @DejanMilicic's comment:
There are 2 new properties in the new EditableMediaUploadAttribute:
* *PreferredSizes*, where you can establish which size is pre-selected (original, wide, thumb,...)  or leave it blank
* *Extensions*, by default images (".jpg,.png,.jpeg,.gif,.bmp), but you can establish the allowed extensions, using a comma separated list, for instance ".pdf" or ".doc,.docx,.xls,.xlsx", etc.

I'll be very glad to bring this new functionalities, interaction and Look&Feel to the File Manager.

**Comment by DejanMilicic** [comment]
@libardo In version 2.9, MediaBrowser is not working. Window opens,  but instead of media browser, following error shows

[InvalidOperationException: The resource class for this page was not found.  Please check if the resource file exists and try again.]
   System.Web.Compilation.LocalResXResourceProvider.CreateResourceManager() +1470809
   System.Web.Compilation.BaseResXResourceProvider.GetObject(String resourceKey, CultureInfo culture) +28
   System.Web.Compilation.ResourceExpressionBuilder.GetResourceObject(IResourceProvider resourceProvider, String resourceKey, CultureInfo culture, Type objType, String propName) +21
   System.Web.UI.TemplateControl.GetLocalResourceObject(String resourceKey) +46
   ASP.n2_content_navigation_mediabrowser_aspx.__Render__control1(HtmlTextWriter __w, Control parameterContainer) in \N2\Content\Navigation\MediaBrowser.aspx:21

**Comment by libardo** [comment]
New version coming up

**Comment by laurachan** [comment]
I just upgraded to version 2.9.0.1. The new media browser (which, by the way, looks great!) works when I use IIS Express, but unfortunately not when I use Local IIS and have the project built into a virtual directory. I am using Chrome 46.
* Clicking on a subdirectory fails to navigate there. A 404 error also comes up in the console for http://localhost/MyVirtualDir/N2/Resources/Css/smoothness/images/animated-overlay.gif 
* If I try to upload a file after that, MediaBrowser.js line 551 logs the following to the console: Object {Status: "Error", Message: "Upload failed"}

**Comment by DejanMilicic** [comment]
Everyone, what do you think about adding some kind of image preview for selected image into property marked by EditableMediaAttribute? For example, small icon (3rd one) which, when hovered would bring tooltip-like preview of selected image.

**Comment by laurachan** [comment]
@DejanMilicic sounds like it could be useful. How would it show for larger image sizes?

**Comment by laurachan** [comment]
I got the following error after trying to set an [EditableMediaBrowser] detail as required:

Unable to cast object of type 'N2.Web.UI.WebControls.SelectingMediaControl' to type 'N2.Web.UI.WebControls.SelectingUploadCompositeControl'.

[InvalidCastException: Unable to cast object of type 'N2.Web.UI.WebControls.SelectingMediaControl' to type 'N2.Web.UI.WebControls.SelectingUploadCompositeControl'.]
   N2.Details.EditableMediaAttribute.AddRequiredFieldValidator(Control container, Control editor) in c:\Work\github\n2cms\src\Framework\N2\Details\EditableMediaAttribute.cs:121
   N2.Details.AbstractEditableAttribute.AddValidation(Control container, Control editor) in c:\Work\github\n2cms\src\Framework\N2\Details\AbstractEditableAttribute.cs:203
   N2.Details.AbstractEditableAttribute.AddTo(Control container) in c:\Work\github\n2cms\src\Framework\N2\Details\AbstractEditableAttribute.cs:195

**Comment by SntsDev** [comment]
@laurachan , 
* I changed the casting class for the Required attribute to work. New version should include it (@libardo).
* The missing image is in the repo, but not included in the ZIP. 
@DejanMilicic, How do you specify that certain files have to be included in the ZIP?
* Was the virtual dir working with the previous tree.aspx browser? 
* I'll modify MediaBrowser.js to show the errors in the UI not only in the console.

**Comment by laurachan** [comment]
Thanks @SntsDev. I can confirm that the old tree.aspx browser did work in a virtual directory.

**Comment by SntsDev** [comment]
@laurachan, so you can see the first list of files (and directories) when opening the Media Browser, but not being able to change dir, can't you?

**Comment by laurachan** [comment]
@SntsDev yes, that's pretty much what happens. Here are the details (with a slight correction to what I posted earlier):
* When I first open the media browser, I do see the first list of files/directories. The path shown below the search bar is [root] / MyVirtualDir / upload. Selecting a file from here works.
* If I click on a subdirectory, it takes me to the root and the 404 I mentioned earlier shows up. From the root, I can only see a directory called upload. Clicking on it goes nowhere.

**Comment by libardo** [comment]
Lots of good ideas and improvements here. Since they are unrelated to the original topic I created some new issues #681, #682, #683, #684, #685. Feel free to add anything I missed.

@SntsDev: Would you be interested in contributing to any of these? Let me know =)

**Comment by DejanMilicic** [comment]
I am reopening this issue since I encountered two more errors.

Existing project is upgraded to latest version of N2. I tested on EditableImageUpload and I get two errors.

1. When I open media browser, "select" button is small rectangle without any text on it, 
2. Selecting desired image via this button is impossible

There are two JS errors in media browser

1. 404 error for : /N2/Resources/Css/smoothness/images/animated-overlay.gif
2. JS error : Cannot read property 'isVisible' of undefined, http://localhost:55095/N2/Resources/Js/MediaSelection.js, line 47

I am testing this on latest version of Chrome, running under IIS Express, VS2015

**Comment by DejanMilicic** [comment]
Also, I just noticed that, when using EditableMediaUpload, uploading of image is not possible (or am I am missing this feature?)

**Comment by DejanMilicic** [comment]
@SntsDev Did you solved some problems I just reported in https://github.com/n2cms/n2cms/pull/687 ?

**Comment by SntsDev** [comment]
@DejanMilicic, 

1. Yes, there is a problem to assign Resources using GetLocalResourceObject. It works on IIS Express but not on IIS. I also tried using Literals and Meta:ResourceKey with default Text values, which half-works because now you can see the labels but I think it's not using the resource values. I'm adding a new Pull Request with these. That is why you see a small blue button without text and no text in the upload-file tab.

2. The 404 error of the overlay.gif is out of my control. The GIF is there but it's not included in the ZIP.

3. The "isVisible" error happens in fast browsers (Webkit). In #687 I fixed it waiting for the onload event. That is why you think you can't select. You can, indeed, and the input value changes, but the window doesn't close. With next Pull Request it will.

4. I'm also adding a better interaction response to files larger than MaxRequestLength to handle it in the browser (not sending those files).
